modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java

```diff
@@ -0,0 +1,692 @@
+            stream = streamSpi.createInputStreamInstance(fileURL, ImageIO.getUseCache(),
+                    ImageIO.getCacheDirectory());
+            readerSpi = getReaderSpiFromStream(suggestedSPI, stream);
+        } catch (Exception e) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
+                LOGGER.log(Level.WARNING,
+                        "Unable to create a Reader for File: " + inputFile.getCanonicalPath(), e);
+            }
+            throw new IllegalArgumentException(e);
+        } finally {
+            if (stream != null) {
+                try {
+                    stream.close();
+                } catch (Exception e) {
+                    if (LOGGER.isLoggable(Level.SEVERE)) {
+                        LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                    }
+                }
+            }
+        }
+        // Handling Overviews
+        File overviewFile = new File(inputFile.getAbsolutePath() + OVR_EXTENSION);
+        hasDatasetLayout = layout != null;
+        if (hasDatasetLayout && layout.getExternalOverviews() != null) {
+            overviewFile = layout.getExternalOverviews();
+        }
+        if (overviewFile.exists() && overviewFile.canRead()) {
+            // Creating overview file URL
+            ovrURL = DataUtilities.fileToURL(overviewFile);
+            // Creating cached SPIs
+            overviewStreamSpi = getInputStreamSPIFromURL(ovrURL);
+            ImageInputStream ovrStream = null;
+            try {
+                ovrStream = overviewStreamSpi.createInputStreamInstance(ovrURL,
+                        ImageIO.getUseCache(), ImageIO.getCacheDirectory());
+                overviewReaderSpi = getReaderSpiFromStream(null, ovrStream);
+            } catch (Exception e) {
+                if (LOGGER.isLoggable(Level.WARNING)) {
+                    LOGGER.log(
+                            Level.WARNING,
+                            "Unable to create a Reader for File: "
+                                    + overviewFile.getCanonicalPath(), e);
+                }
+                throw new IllegalArgumentException(e);
+            } finally {
+                if (ovrStream != null) {
+                    try {
+                        ovrStream.close();
+                    } catch (Exception e) {
+                        if (LOGGER.isLoggable(Level.SEVERE)) {
+                            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                        }
+                    }
+                }
+            }
+        } else {
+            // No Overview file so we fall back to the original file spis
+            overviewStreamSpi = streamSpi;
+            overviewReaderSpi = readerSpi;
+        }
+
+        // Getting number of Overviews
+        int numOverviews = 0;
+        if (hasDatasetLayout) {
+            numInternalOverviews = layout.getNumInternalOverviews();
+            numExternalOverviews = layout.getNumExternalOverviews() > 0 ? layout
```

Comment:
This is not documented here: https://github.com/geosolutions-it/imageio-ext/blob/master/library/geocore/src/main/java/it/geosolutions/imageio/maskband/DatasetLayout.java